### PR TITLE
Creating ask-sdk-model-runtime package

### PR DESCRIPTION
**Pull request by Shreyas-vgr**
## Description

- Created an ask-sdk-model-runtime package which contains the defintions of LWA servies, implementations of ApiClient and its componenets, also default Serializer definitions.
- This package can be reused across multiple models released by ask sdk team in future, also the current ask-sdk-models (python) can add this as dependency.

## Testing
ask-sdk-model-runtime package has unit test cases attached for all the components defined.

## Checklist
- [x] My code follows the code style of this project
- [x] My change requires a change to the documentation
- [x] The change is generic and can be done across all model classes.
- [x] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ask-sdk-model-runtime/README.rst [hunk 1]
```diff
@@ -0,0 +1,52 @@
+====================================================
+ASK SDK Model Runtime - Base components of SMAPI SDK
+====================================================
+
+ask-sdk-model-runtime is the package for Alexa Skills Kit (ASK) by
+the Software Development Kit (SDK) team for Python. It provides the
+base components and default implementations which work as the boiler
+plate code for developing Alexa Skills.
```
Comment (nikhilym) [inline ask-sdk-model-runtime/README.rst]: We can provide a little bit more information on what this package is about. For ex, something like 

````
ask-sdk-model-runtime is the common models package that provides common implementations like Serializer, ApiClient etc. that is used by the SDK models.
````

File: ask-sdk-model-runtime/ask_sdk_model_runtime/__version__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS
+# OF ANY KIND, either express or implied. See the License for the
+# specific language governing permissions and limitations under the
+# License.
+#
+
+__pip_package_name__ = 'ask-sdk-model-runtime'
+__description__ = ('The ASK SDK Model Runtime package provides common '
+                   'functionalities, for building Alexa Skills.')
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/__version__.py]: Similar here

File: ask-sdk-model-runtime/ask_sdk_model_runtime/__version__.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+__url__ = 'https://github.com/alexa/alexa-skills-kit-sdk-for-python'
+__version__ = '0.1'
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/__version__.py]: Change version here as well

File: ask-sdk-model-runtime/ask_sdk_model_runtime/base_service_client.py [hunk 4]
```diff
@@ -42,15 +42,15 @@ class BaseServiceClient(object):
     instantiation is not supported
 
     :param api_configuration: ApiConfiguration implementation
-    :type api_configuration: ask_sdk_model.services.api_configuration.ApiConfiguration
+    :type api_configuration: ask_sdk_model.api_configuration.ApiConfiguration
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/base_service_client.py]: typo: `ask_sdk_model_runtime`

File: ask-sdk-model-runtime/ask_sdk_model_runtime/base_service_client.py [hunk 5]
```diff
@@ -215,7 +215,7 @@ def __build_query_string(query_params, is_query_start):
             encoded_query_params.append(
                 (param_name_encoded, param_value_encoded))
 
-        return query_string + urlencode(encoded_query_params, doseq=True)
+        return query_string + urlencode(encoded_query_params)
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/base_service_client.py]: Why was this removed? We need the `doseq=True` for supporting multiple query params for same key

File: ask-sdk-model-runtime/ask_sdk_model_runtime/exceptions.py [hunk 6]
```diff
@@ -31,10 +35,15 @@ class ServiceException(Exception):
...
     """
 
-    def __init__(self, message, status_code, headers, body):
-        # type: (str, int, Optional[List[Tuple[str, str]]], Optional[T]) -> None
+    def __init__(self, message,  # type: str
+                 status_code,  # type: int
+                 headers,  # type: Optional[List[Tuple[str, str]]]
+                 body  # type: Optional[T]
+                 ):
+        # type: (...) -> None
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/exceptions.py]: Although I like this formatting of mentioning the type hints for each constructor parameter, let's stick to the current format we have. I think the last time I tried this, on some python versions, the IDE didn't recognize the type hints.

File: ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/access_token.py [hunk 7]
```diff
@@ -20,11 +20,12 @@
 if typing.TYPE_CHECKING:
     from datetime import datetime
 
+
 class AccessToken(object):
     """Represents the access token provided by LWA (Login With Amazon).
 
     This is a wrapper class over
-    :py:class:`ask_sdk_model.services.lwa.access_token_response.AccessTokenResponse`
+    :py:class:`ask_sdk_model_runtime.services.lwa.access_token_response.AccessTokenResponse`
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/access_token.py]: typo: `ask_sdk_model_runtime.lwa.access_token_response.AccessTokenResponse`

File: ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/access_token_request.py [hunk 8]
```diff
@@ -34,25 +34,23 @@ class AccessTokenRequest(object):
...
+    attribute_map = {'client_id': 'client_id',
+        'client_secret': 'client_secret', 'refresh_token': 'refresh_token',
+        'scope': 'scope'}
+
+    def __init__(self, client_id=None,  # type: Optional[str]
+                 client_secret=None,  # type: Optional[str]
+                 scope=None,  # type: Optional[str]
+                 refresh_token=None  # type: Optional[str]
+                 ):
+        # type: (...) -> None
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/access_token_request.py]: Same for the type hints here, as described above.

File: ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/lwa_client.py [hunk 9]
```diff
@@ -39,7 +39,7 @@ class LwaClient(BaseServiceClient):
         Serializer and ApiClient. The authorization value and api endpoint
         is not used by the LWA Client.
     :type api_configuration:
-        ask_sdk_model.services.api_configuration.ApiConfiguration
+        ask_sdk_model_runtime.services.api_configuration.ApiConfiguration
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/lwa_client.py]: typo: wrong namespace across the changes in the file

File: ask-sdk-model-runtime/ask_sdk_model_runtime/api_client.py [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+from .api_client_response import ApiClientResponse
+
+from .exceptions import ApiClientException
+
+if typing.TYPE_CHECKING:
+    from typing import Callable, Dict, List, Tuple
+    from ask_sdk_model_runtime import ApiClientRequest
+
+
+class ApiClient(object):
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/api_client.py]: Can we rename this back to `DefaultApiClient` and also have this subclassed from `ApiClient`, so that we can keep it backwards compatible.

File: ask-sdk-model-runtime/ask_sdk_model_runtime/serializer.py [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+    long  # type: ignore
+except NameError:
+    long = int
+
+if typing.TYPE_CHECKING:
+    from typing import TypeVar, Dict, List, Tuple, Union, Any
+    T = TypeVar('T')
+
+
+class Serializer(object):
```
Comment (nikhilym) [inline ask-sdk-model-runtime/ask_sdk_model_runtime/serializer.py]: Similar to `ApiClient`, let's call this `DefaultSerializer` subclasses from the abstract `Serializer` class.

File: ask-sdk-model/ask_sdk_model/services/proactive_events/proactive_events_service_client.py [hunk 12]
```diff
@@ -25,7 +25,7 @@
 from ask_sdk_model.services.api_response import ApiResponse
 
 from ask_sdk_model.services.authentication_configuration import AuthenticationConfiguration
```
Comment (nikhilym) [inline ask-sdk-model/ask_sdk_model/services/proactive_events/proactive_events_service_client.py]: These imports and the other service imports in the class needs to be changed. But won't these be updated during the model generation? Similarly for `skill_messaging_service_client.py`.

File: ask-sdk-model-runtime/CHANGELOG.rst [hunk 13]
```diff
@@ -0,0 +1,8 @@
+=========
+CHANGELOG
+=========
+
+0.1
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/CHANGELOG.rst]: Yes, will update it to 1.0

File: ask-sdk-model-runtime/MANIFEST.in [hunk 14]
```diff
@@ -0,0 +1,5 @@
+include README.rst
+include CHANGELOG.rst
+include LICENSE
+include requirements.txt
+recursive-exclude tests *
```
[the remark this answers is not in the document]
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/MANIFEST.in]: I will add recursive include ask-sdk-model-runtime py.typed explicity as mentioned [here](https://github.com/alexa/alexa-apis-for-python/blob/master/ask-sdk-model/MANIFEST.in#L5)

File: ask-sdk-model-runtime/README.rst (hunk 1 above)
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/README.rst]: Ok

File: ask-sdk-model-runtime/ask_sdk_model_runtime/__version__.py (hunk 2 above)
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/ask_sdk_model_runtime/__version__.py]: Will update

File: ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/access_token_request.py (hunk 8 above)
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/ask_sdk_model_runtime/lwa/access_token_request.py]: Ok, will revert it back to inline type comments

File: ask-sdk-model-runtime/ask_sdk_model_runtime/exceptions.py (hunk 6 above)
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/ask_sdk_model_runtime/exceptions.py]: Sure, will revert it back to old format

File: ask-sdk-model-runtime/ask_sdk_model_runtime/api_client.py (hunk 10 above)
Comment (Shreyas-vgr) [reply ask-sdk-model-runtime/ask_sdk_model_runtime/api_client.py]: Sure, will add ApiClient abstract definition and implement `DefaultApiClient` and also do the same with `Serializer`.

File: ask-sdk-model/ask_sdk_model/services/proactive_events/proactive_events_service_client.py (hunk 12 above)
Comment (Shreyas-vgr) [reply ask-sdk-model/ask_sdk_model/services/proactive_events/proactive_events_service_client.py]: Yes, will update them for now, yes they will be updated during the model generation.
